### PR TITLE
Drop rustc-serialize dependency

### LINKED ISSUE #583: Migrate from rustc-serialize to Serde
It looks like it went well in [cargo](https://github.com/rust-lang/cargo/pull/3682), at some point down the line we should do migrate as well.

### PR DESCRIPTION
**Pull request by ordian**
~~~Breaking change (but who cares?).~~~
- ~~~encodable* types were renamed to serializable* for consistency with serde~~~
- error messages were tweaked a little bit

~~~TODO (migrate in dependencies):~~~
- [x] ~~~update clippy~~~
- [ ] [mime-types](https://github.com/conduit-rust/mime-types) ([#8](https://github.com/conduit-rust/mime-types/pull/8))
- [ ] [conduit-cookie](https://github.com/conduit-rust/conduit-cookie) ([#3](https://github.com/conduit-rust/conduit-cookie/pull/3), [#4](https://github.com/conduit-rust/conduit-cookie/pull/4))
- [ ] [conduit-json-parser](https://github.com/conduit-rust/conduit-json-parser) ([#3](https://github.com/conduit-rust/conduit-json-parser/pull/3))  

Closes #583.

### CODE REVIEW

**Comment by ordian** [comment]
So this is ready to merge. 

It doesn't drop rustc-serialize in transitive dependencies yet, but this can be done in a separate PR.  
To do that we need a version bump from mime-types, conduit-cookie, conduit-json-parser, also
conduit-static needs to be updated on a new mime-types and bump it's version as well (or we can switch from conduit to other web application interface). 

cc @carols10cents what do you think?

File: .travis.yml [hunk 1]
```diff
@@ -49,7 +49,7 @@ matrix:
     - cargo build
     - cargo test
     - yarn test
-  - rust: nightly-2017-03-04
+  - rust: nightly-2017-06-29
```
Comment (sgrif) [inline .travis.yml]: This seems unrelated to the other changes in this PR

**Comment by sgrif** [comment]
Is the rename of `encodable` to `serializable` really necessary? It seems like it just adds a ton of churn for very little gain. After switching to Serde I suspect we can just eliminate a lot of these anyway.

File: .travis.yml (hunk 1 above)
Comment (ordian) [inline .travis.yml]: This is related to updating clippy. Old clippy (0.0.118) depended on rustc-serialize and new one (0.0.141) won't build with nightly-2017-03-04.

**Comment by ordian** [comment]
> Is the rename of encodable to serializable really necessary? It seems like it just adds a ton of churn for very little gain. After switching to Serde I suspect we can just eliminate a lot of these anyway.

Good point. I reverted this change. 

I would prefer eliminating encodable structs in a separate pr, because currently it is blocked on postgres release [with chrono 0.4](https://github.com/sfackler/rust-postgres/pull/272).

**Comment by sgrif** [comment]
Yes, I agree that the elimination should not be part of this PR>

File: src/upload.rs [hunk 2]
```diff
@@ -74,48 +79,56 @@ where
...
+impl<'de> Deserialize<'de> for CrateVersion {
+    fn deserialize<D: Deserializer<'de>>(d: D) -> Result<CrateVersion, D::Error> {
+        let s = String::deserialize(d)?;
         match semver::Version::parse(&s) {
             Ok(v) => Ok(CrateVersion(v)),
-            Err(..) => Err(d.error(&format!("invalid semver: {}", s))),
+            Err(..) => {
+                let value = de::Unexpected::Str(&s);
+                let expected = "a valid semver";
+                Err(de::Error::invalid_value(value, &expected))
```
Comment (sgrif) [inline src/upload.rs]: The semver crate has a `serde` feature. Should we just use that instead?

**Comment by sgrif** [comment]
I'm a little worried about whether this changes our response structure in subtle ways that will break the front-end and won't be caught by our tests, but overall this seems good.

File: src/upload.rs (hunk 2 above)
Comment (ordian) [inline src/upload.rs]: I tried to at first, but failed. Because of conduit, we must use semver 0.5, which doesn't have serde feature.

**Comment by ordian** [comment]
> I'm a little worried about whether this changes our response structure in subtle ways that will break the front-end and won't be caught by our tests

This PR doesn't change json structure, only the encoder/decoder is changed. Assuming serde_json is correct, it won't break front-end.

**Comment by carols10cents** [comment]
Nice!! Thank you for this!!! Love how it gets rid of things like the `fixup` function and just makes it a serde rename 👍 💯 

> also
> conduit-static needs to be updated on a new mime-types and bump its version as well (or we can switch from conduit to other web application interface).

Bumping the versions of anything out of date sounds great-- I'd rather do that than switch to another web app framework right now. I'm not *opposed* to switching from conduit to something else, it'd just be a bigger project, a larger change that can't easily be made incrementally, and that kind of thing makes me nervous :) If switching frameworks makes the code easier to understand and possibly more familiar for contributors, I'm up for being convinced that it's worth the tradeoff though :)

The error messages printed through cargo when a publish gets rejected are a little bit uglier. Before this change:

```
error: api errors: invalid upload request: ApplicationError("invalid crate name specified: aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa. Valid crate names must start with a letter; contain only letters, numbers, hyphens, or underscores; and have 64 or fewer characters.")
```

after: 

```
error: api errors: invalid upload request: ErrorImpl { code: Message("invalid value: string \"aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa\", expected a valid crate name to start with a letter, contain only letters, numbers, hyphens, or underscores and have at most 64 characters"), line: 1, column: 94 }
```

Just that extra `ErrorImpl { code: Message(` where it only said `ApplicationError` before. I'm going to merge this anyway since people publishing crates should be fine reading Rust errors, just something to think about and maybe fix if folks complain. Not sure if the fix to this would be in crates.io or in cargo...

**Comment by ordian** [comment]
> I'm not opposed to switching from conduit to something else, it'd just be a bigger project, a larger change that can't easily be made incrementally, and that kind of thing makes me nervous :) 

I understand the concern.

> If switching frameworks makes the code easier to understand and possibly more familiar for contributors, I'm up for being convinced that it's worth the tradeoff though :)

It probably would, since conduit is [not maintained](https://github.com/conduit-rust/conduit/pulse/monthly). But it's better to wait for a framework with a mature http2 and async support (hyper is anync, but not a framework, iron is not async yet).

> Just that extra ErrorImpl { code: Message( where it only said ApplicationError before.

Does [this PR](https://github.com/rust-lang/crates.io/pull/860) fix the issue?
